### ISSUE TITLE
Run saga policy (prevent saga duplication in more flexible way)

### ISSUE THREAD

**Issue by afitiskin** [react-boilerplate/react-boilerplate#1183]
**Feature request**

Hello guys. This topic is not new. I've checked issues #390, #637 and #888 and have a few ideas how to improve proposed solutions.

So, I like the idea and solution of preventing saga duplication proposed by @kraffslol at #390 (by changing saga injector), but this solution was rejected. The point is:
> Sagas themselves should manage their lifecycle 

and proposal solution is to cancel watchers on `LOCATION_CHANGE` action:
```javascript
yield take(LOCATION_CHANGE);
yield cancel(watcher);
```

But in some cases this way has problems. #888 is an example when new saga should not be run automatically, because we would like to handle this kind of location change action in previously run saga. To do that we have to tweak async injector (in other case we well have saga duplicates that will ruin our application).

Long story short, what about to have a combined solution? To do that we can add some kind of saga run policy. Injector should check this policy and run (or not run) each saga. This way has benefits of both solutions and can handle all possible saga-usage cases.

Code example to illustrate my idea:
```javascript
// some container's sagas.js

export function* daemonSaga() {
  // run only once on load
}

export function* disposableSaga() {
  // run on each page enter
}
disposableSaga.runAlways = true;

export default [
  daemonSaga,
  disposableSaga,
];

// store.js
// ... much code here
store.asyncReducers = {};
store.asyncSagas = {};
return store;

// asyncInjectors.js
export function injectAsyncSagas(store, isValid) {
  return function injectSagas(sagas) {
    // invariant and warning here...
  
    sagas.forEach((saga) => {
      if (!Reflect.has(store.asyncSagas, saga) || saga.runAlways === true) {
        store.runSaga(saga);
        store.asyncSagas[saga] = true;
      }
    });
  }
}
```

What do you think?

**Comment by nhducit** [comment]
I think even if your implementation, we still have the #1224 issue

**Comment by afitiskin** [comment]
I changed a bit my proposal solution in #1315 to have backward compatibility. By default all sagas lifecycle is exactly as before. If you want to get rid of duplicates for some sagas (without manual watching for `LOCATION_CHANGE` and `cancel`) you just need to set saga's `isDaemon` attribute to `true` and that's it!

Copy of example from PR:
Before:
```javascript
/**
 * Watches for LOAD_REPOS actions and calls getRepos when one comes in.
 * By using `takeLatest` only the result of the latest API call is applied.
 */
export function* getReposWatcher() {
  yield fork(takeLatest, LOAD_REPOS, getRepos);
}

/**
 * Root saga manages watcher lifecycle
 */
export function* githubData() {
  // Fork watcher so we can continue execution
  const watcher = yield fork(getReposWatcher);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [
  githubData,
];
```

After:
```javascript
/**
 * Watches for LOAD_REPOS actions and calls getRepos when one comes in.
 * By using `takeLatest` only the result of the latest API call is applied.
 */
export function* getReposWatcher() {
  yield fork(takeLatest, LOAD_REPOS, getRepos);
}
getReposWatcher.isDaemon = true;

// Bootstrap sagas
export default [
  getReposWatcher,
];
```

**Comment by justingreenberg** [comment]
this should be resolved by #1746 — feedback is welcome!
